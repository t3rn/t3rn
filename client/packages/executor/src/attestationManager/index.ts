import { Sdk } from "@t3rn/sdk";
import { config } from "../../config/config";
import fs from "fs";
import Web3, { Contract } from "web3";
import { logger } from "../../src/logging";
import { Batch, ConfirmationBatch } from "./batch";
import { ethers } from "ethers";
import { Prometheus } from "../prometheus";

interface IBatch {
  nextCommittee: string[];
  bannedCommittee: string[];
  committedSfx: string[];
  revertedSfx: string[];
  index: number;
  expectedBatchHash: string;
  signatures: [string, string][];
  created: number;
}

/**

 * @group Attestions
 */
export class AttestationManager {
  batches: ConfirmationBatch[] = [];
  rpc: string;
  receiveAttestationBatchContract: Contract<never>;
  web3: Web3;
  wallet: ReturnType<typeof this.web3.eth.accounts.privateKeyToAccount>;
  currentCommitteeSize = 0;
  currentBatchIndex = 0;
  currentCommitteeTransitionCount = 0;
  prometheus: Prometheus;

  constructor(public client: Sdk["client"], prometheus: Prometheus) {
    if (config.attestations.ethereum.privateKey === undefined) {
      throw new Error("Ethereum private key is not defined");
    }

    this.client = client;
    this.rpc = config.attestations.ethereum.rpc;
    this.web3 = new Web3(this.rpc);

    this.wallet = this.web3.eth.accounts.privateKeyToAccount(
      config.attestations.ethereum.privateKey
    );

    if (this.wallet.address === undefined) {
      throw new Error("Ethereum wallet address is not defined");
    }

    logger.info("Wallet address: " + this.wallet.address);

    const receiveAttestationBatchAbi = JSON.parse(
      fs.readFileSync(
        "./src/attestationManager/contracts/AttestationsVerifier.abi.json",
        "utf8"
      )
    );
    const receiveAttestationBatchAddress =
      config.attestations.ethereum.attestationVerifierAddress;
    this.receiveAttestationBatchContract = new this.web3.eth.Contract(
      receiveAttestationBatchAbi.abi,
      receiveAttestationBatchAddress
    );
    this.prometheus = prometheus;
  }

  batchEncodePacked(batch: Batch) {
    return ethers.solidityPacked(
      ["address[]", "address[]", "bytes32[]", "bytes32[]", "uint32"],
      [
        batch.nextCommittee,
        batch.bannedCommittee,
        batch.committedSfx,
        batch.revertedSfx,
        batch.index,
      ]
    );
  }

  async fetchBatches() {
    logger.info("Fetching batches from chain...");
    const attesters = this.client.query.attesters;
    const rawBatches = await attesters.batches(
      config.attestations.ethereum.name
    );
    const fetchedData = rawBatches.toJSON() as unknown as Array<IBatch>;
    this.batches = fetchedData.map((batch) => {
      return {
        nextCommittee: batch.nextCommittee || [],
        bannedCommittee: batch.bannedCommittee || [],
        committedSfx: batch.committedSfx || [],
        revertedSfx: batch.revertedSfx || [],
        index: batch.index,
        signatures: batch.signatures.map(([, signature]) => signature),
        created: batch.created,
      } as ConfirmationBatch;
    });
    logger.info("We have " + fetchedData.length + " batches pending");
    this.prometheus.attestationsBatchesPending.set(fetchedData.length);
  }

  async listener() {
    logger.info("Listening for NewConfirmationBatch events...");
    // Subscribe to all events and filter based on the specified event method
    this.client.query.system.events(async (events: any) => {
      if (events === undefined || events === null) {
        return;
      }

      logger.debug({ events_count: events.length }, `Received events`);

      const attesterEvents = events
        .toHuman()
        .filter((event) => event.event.section == "attesters");

      await this.processConfirmedBatches(attesterEvents);
    });
  }
  private async processConfirmedBatches(events: any) {
    // Loop through the Vec<EventRecord>
    await Promise.all(
      events.map(async (record) => {
        // Extract the phase, event and the event types
        const { event } = record;

        logger.debug(
          {
            event: event.method,
            data: event.data,
          },
          "Received attesters event"
        );
        this.prometheus.attestationsEvents.inc({ method: event.method });
        if (record.event.method != "NewConfirmationBatch") {
          return;
        }

        const batch = event.data as ConfirmationBatch;
        const messageHash = this.getMessageHash(batch);

        await this.receiveAttestationBatchCall(batch, messageHash);
      })
    );
  }

  getMessageHash(batch: Batch) {
    const encodedBatch = this.batchEncodePacked(batch);
    const messageHash = ethers.keccak256(encodedBatch);
    return messageHash;
  }

  async getBlockHash(blockNumber: number): Promise<string> {
    const blockHash = await this.client.rpc.chain.getBlockHash(blockNumber);
    return blockHash.toHex();
  }

  // TODO: not used
  getMessageHashWithPrefix(batch: Batch) {
    const prefix = ethers.hexlify(
      ethers.toUtf8Bytes("\x19Ethereum Signed Message:\n32")
    );
    const encodedBatch = ethers.keccak256(this.batchEncodePacked(batch));
    const messageHash = ethers.keccak256(ethers.concat([prefix, encodedBatch]));
    logger.debug("MsgHash with prefix: " + messageHash);

    return messageHash;
  }

  async processAttestationBatches() {
    await this.fetchAttesterContractData();
    await this.fetchBatches();

    if (this.batches.length > this.currentBatchIndex) {
      logger.info(
        `We have ${
          this.batches.length - this.currentBatchIndex
        } pending batches to process`
      );
      this.processPendingAttestationBatches();
    }

    await this.listener();
  }

  async processPendingAttestationBatches() {
    for (const batch of this.batches.slice(
      Number(this.currentBatchIndex) + 1
    )) {
      logger.info(
        `Batch ${
          batch.index
        } processing, created at block ${await this.getBlockHash(
          batch.created
        )}`
      );

      const messageHash = this.getMessageHash(batch);
      logger.debug({ batch, messageHash }, `Batch data`);

      await this.receiveAttestationBatchCall(batch, messageHash);
    }
  }

  async receiveAttestationBatchCall(
    batch: ConfirmationBatch,
    messageHash: string
  ) {
    const contractMethod =
      this.receiveAttestationBatchContract.methods.receiveAttestationBatch(
        // @ts-ignore - ethers.js types are not up to date
        batch.nextCommittee,
        batch.bannedCommittee,
        batch.committedSfx,
        batch.revertedSfx,
        batch.index,
        messageHash,
        batch.signatures
      );

    const encodedABI = contractMethod.encodeABI();

    const gasPrice = await this.web3.eth.getGasPrice();
    const estimatedGas = await contractMethod.estimateGas({
      from: this.wallet.address,
    });

    const transactionObject = {
      to: this.receiveAttestationBatchContract.options.address,
      from: this.wallet.address,
      data: encodedABI,
      gas: 500000,
      gasPrice: gasPrice,
      estimatedGas: estimatedGas,
    };

    const signedTransaction = await this.wallet.signTransaction(
      transactionObject
    );

    try {
      const transactionReceipt = await this.web3.eth.sendSignedTransaction(
        signedTransaction.rawTransaction
      );
      logger.info(
        { receipt: transactionReceipt },
        `Batch ${batch.index} procesed!`
      );
      this.prometheus.attestatonsBatchesProcessed.inc();
    } catch (error) {
      logger.warn({ error: error }, "Error sending transaction: ");
      this.prometheus.attestatonsBatchesFailed.inc();
      // throw new Error("Error sending transaction: " + error);
    }
  }

  private async fetchAttesterContractData() {
    this.currentCommitteeSize = Number(
      await this.receiveAttestationBatchContract.methods.committeeSize().call()
    );
    this.prometheus.attestationVerifierCurrentCommitteeSize.set(
      this.currentCommitteeSize
    );

    this.currentBatchIndex = Number(
      await this.receiveAttestationBatchContract.methods
        .currentBatchIndex()
        .call()
    );
    this.prometheus.attestationVerifierCurrentBatchIndex.set(
      this.currentBatchIndex
    );

    this.currentCommitteeTransitionCount = Number(
      await this.receiveAttestationBatchContract.methods
        .currentCommitteeTransitionCount()
        .call()
    );
    this.prometheus.attestationVerifierCurrentCommitteeTransitionCount.set(
      this.currentCommitteeTransitionCount
    );

    logger.debug(
      {
        currentCommitteeSize: this.currentCommitteeSize,
        currentBatchIndex: this.currentBatchIndex,
        currentCommitteeTransitionCount: this.currentCommitteeTransitionCount,
      },
      "Etherum Contract State"
    );
  }
}
