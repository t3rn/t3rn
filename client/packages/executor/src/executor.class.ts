import { Sdk } from "@t3rn/sdk";
import { Keyring } from "@polkadot/api";
import { cryptoWaitReady } from "@polkadot/util-crypto";
import { KeyringPair } from "@polkadot/keyring/types";
import { ExecutionManager } from "./executionManager";
import { config, Config } from "../config/config";
import { getBalanceWithDecimals } from "./utils";
import { Logger } from "pino";
import { Prometheus } from "./prometheus";
import { logger } from "./logging";

/** An executor instance. */
export class Executor {
  name: string;
  circuitClient: Sdk["client"];
  executionManager: ExecutionManager;
  sdk: Sdk;
  signer: KeyringPair;
  config: Config;
  logger: Logger;
  prometheus: Prometheus;

  /**
   * Initializes an executor instance.
   *
   * @param name Display name and config identifier for an instance
   */
  constructor(prometheus: Prometheus) {
    this.prometheus = prometheus;
    this.config = config;
  }

  /**
   * Sets up and configures an executor instance.
   */
  async setup(): Promise<Executor> {
    await cryptoWaitReady();

    if (!this.config.circuit.signerKey) {
      logger.error("Missing circuit signer key");
      throw Error;
    }

    this.signer = new Keyring({ type: "sr25519" }).addFromMnemonic(
      this.config.circuit.signerKey,
    );
    this.sdk = new Sdk(this.config.circuit.rpc, this.signer);
    this.circuitClient = await this.sdk.init();

    // Set nonce on start
    this.sdk.nonce = (
      await this.circuitClient.rpc.system.accountNextIndex(this.signer.address)
    ).toNumber();
    logger.info(`Set nonce for circuit to ${this.sdk.nonce}`);

    // TODO: print wallet balance on available networks
    const balance = await getBalanceWithDecimals(
      this.circuitClient,
      this.signer.address,
    );

    // Convert the balance to a human-readable format
    logger.info(
      {
        circuit_signer_address: this.signer.address,
        circuit_signer_balance: balance,
      },
      `Circuit Signer Address`,
    );
    this.prometheus.executorBalance.set(
      {
        signer: this.signer.address,
        target: this.config.circuit.name,
      },
      balance,
    );

    this.executionManager = new ExecutionManager(
      this.circuitClient,
      this.sdk,
      this.logger,
      this.config,
      this.prometheus,
    );
    await this.executionManager.setup(
      this.config.gateways,
      this.config.vendors,
    );

    // In case anything goes wrong with nonces we should reset it on each new block to chain state
    this.watchBlock();

    logger.info("Executor setup complete");
    return this;
  }

  async watchBlock() {
    await this.circuitClient.rpc.chain.subscribeNewHeads(async () => {
      // Process the new block header
      this.sdk.nonce = (
        await this.circuitClient.rpc.system.accountNextIndex(
          this.signer.address,
        )
      ).toNumber();
      logger.info(`Set nonce for circuit to ${this.sdk.nonce}`);
    });
  }
}
