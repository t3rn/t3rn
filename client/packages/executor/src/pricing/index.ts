import { CoingeckoPricing } from "./coingecko"
import { BehaviorSubject } from "rxjs"

/**
 * Class used for keeping track of asset prices. This class has access to different price sources, allowing the generation of averages of
 * any kind. All values are subjects, allowing the user to subscribe to price updates.
 *
 * @group Pricing
 */
export class PriceEngine {
    coingecko: CoingeckoPricing

    constructor() {
        this.coingecko = new CoingeckoPricing()
    }

    /**
     * Returns the price of an asset in USD
     *
     * @param assetId Ticker of the asset we want the price of. These are set in the config file.
     * @returns The price of the asset in USD as a subject
     */
    getAssetPrice(assetId: string): BehaviorSubject<number> {
        return this.coingecko.prices[assetId]
    }
}

export { CoingeckoPricing }
