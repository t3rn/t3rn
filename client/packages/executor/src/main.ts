import { AttestationManager } from "./attestationManager";
import { Instance } from "./index";
import { config } from "../config/config";
import { logger } from "./logging";

async function main() {
  logger.info("Starting executor");
  const instance = new Instance(process.env.EXECUTOR);
  await instance.setup();

  const attestationManager = new AttestationManager(instance.circuitClient);
  if (config.attestations.processPendingBatches) {
    logger.info("Processing Pending Attestation Batches");
    await attestationManager.processPendingAttestationBatches();
  }
  // await attestationManager.listener()
}

main();
