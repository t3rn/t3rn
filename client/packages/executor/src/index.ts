import "@polkadot/api-augment";
import * as dotenv from "dotenv";
import { Sdk } from "@t3rn/sdk";
import { Keyring } from "@polkadot/api";
import { cryptoWaitReady } from "@polkadot/util-crypto";
import { KeyringPair } from "@polkadot/keyring/types";
import { PathLike, existsSync } from "fs";
import { readFile, writeFile, mkdir } from "fs/promises";
import { join } from "path";
import { homedir } from "os";
import "@t3rn/types";
import {
  SubstrateRelayer,
  CostEstimator,
  Estimator,
  Estimate,
  InclusionProof,
} from "./gateways/substrate/relayer";
import { ExecutionManager, PersistedState, Queue } from "./executionManager";
import { Config, Gateway, Circuit, Strategy } from "../config/config";
import { BiddingEngine, BiddingStrategy } from "./bidding";
import { PriceEngine, CoingeckoPricing } from "./pricing";
import { StrategyEngine, SfxStrategy, XtxStrategy } from "./strategy";
import {
  SideEffect,
  Notification,
  NotificationType,
  TxOutput,
  TxStatus,
} from "./executionManager/sideEffect";
import { Execution } from "./executionManager/execution";
import {
  CircuitListener,
  ListenerEvents,
  ListenerEventData,
} from "./circuit/listener";
import { CircuitRelayer } from "./circuit/relayer";
import * as defaultConfig from "../config.json";
import { problySubstrateSeed, createLogger } from "./utils";
import { Logger } from "pino";
import { Prometheus } from "./prometheus";
import { logger } from "./logging";

dotenv.config();

/** An executor instance. */
class Instance {
  name: string;
  circuitClient: Sdk["client"];
  executionManager: ExecutionManager;
  sdk: Sdk;
  signer: KeyringPair;
  config: Config;
  logger: Logger;
  baseDir: PathLike;
  logsDir: undefined | PathLike;
  configFile: PathLike;
  stateFile: PathLike;
  prometheus: Prometheus;

  /**
   * Initializes an executor instance.
   *
   * @param name Display name and config identifier for an instance
   * @param logToDisk Write logs to disk within ~/.t3rn-executor-${name}/logs
   */
  constructor(name = "example", logToDisk = false, prometheus: Prometheus) {
    this.name = name;
    this.baseDir = join(homedir(), `.t3rn-executor-${name}`);
    this.logsDir = logToDisk
      ? join(this.baseDir.toString(), "logs")
      : undefined;
    this.stateFile = join(this.baseDir.toString(), "state.json");
    this.configFile = join(this.baseDir.toString(), "config.json");
    this.prometheus = prometheus;
  }

  /**
   * Sets up and configures an executor instance.
   */
  async setup(): Promise<Instance> {
    await this.configureLogging();
    await this.loadConfig();
    await cryptoWaitReady();

    if (!this.config.circuit.signerKey) {
      throw Error("Instance::setup: missing circuit signer key");
    }

    this.signer = new Keyring({ type: "sr25519" }).addFromSeed(
      Uint8Array.from(
        Buffer.from(this.config.circuit.signerKey.slice(2), "hex"),
      ),
    );
    this.sdk = new Sdk(this.config.circuit.rpc, this.signer);
    this.circuitClient = await this.sdk.init();
    this.circuitClient.on("disconnected", () => {
      this.prometheus.circuitDisconnects.inc({
        endpoint: this.config.circuit.rpc,
      });
    });
    this.executionManager = new ExecutionManager(
      this.circuitClient,
      this.sdk,
      this.logger,
      this.config,
      this.prometheus,
    );
    this.injectState();
    await this.executionManager.setup(
      this.config.gateways,
      this.config.vendors,
    );
    // TODO: on nodejs it just freeze execution
    // this.registerExitListener();
    this.registerStateListener();
    logger.info("Executor setup complete");
    return this;
  }

  /**
   * Loads an instance's config file thereby updating any config changes
   * staged at ../config.json by persisting the effective instance config
   * at ~/.t3rn-executor-${name}/config.json.
   */
  async loadConfig(): Promise<Config> {
    await mkdir(this.baseDir, { recursive: true });
    const persistedConfig = await readFile(this.configFile)
      // if the persisted config file does not exist yet we wanna
      // handle it gracefully because it is probly an initial run
      .then((buf) => {
        try {
          return JSON.parse(buf.toString());
        } catch (err) {
          this.logger.warn(`failed reading ${this.configFile} ${err}`);
          return {};
        }
      })
      .catch((err) => {
        this.logger.warn(`failed reading ${this.configFile} ${err}`);
        return {};
      });
    const config = { ...defaultConfig, ...persistedConfig, name: this.name };
    await writeFile(this.configFile, JSON.stringify(config));
    if (!config.circuit.signerKey?.startsWith("0x")) {
      config.circuit.signerKey = process.env.CIRCUIT_SIGNER_KEY as string;
    }
    config.gateways.forEach((gateway: Gateway) => {
      if (
        gateway.signerKey !== undefined &&
        !problySubstrateSeed(gateway.signerKey)
      ) {
        gateway.signerKey = process.env[
          `${gateway.id.toUpperCase()}_GATEWAY_SIGNER_KEY`
        ] as string;
      }
    });
    if (!problySubstrateSeed(config.circuit.signerKey)) {
      throw Error("Instance::loadConfig: missing circuit signer key");
    }
    if (
      !config.gateways.some((gateway: Gateway) =>
        problySubstrateSeed(gateway.signerKey as string),
      )
    ) {
      throw Error("Instance::loadConfig: missing gateway signer key");
    }
    this.config = config;
    return config;
  }

  /** Loads persisted execution state. */
  async injectState(): Promise<Instance> {
    if (existsSync(this.stateFile)) {
      const state: PersistedState = await readFile(this.stateFile, "utf8").then(
        JSON.parse,
      );
      this.executionManager.inject(state);
    }
    return this;
  }

  /** Configures the instance's pino logger. */
  async configureLogging(): Promise<Instance> {
    if (this.logsDir) {
      await mkdir(this.logsDir, { recursive: true });
      this.logger = createLogger(this.name, this.logsDir.toString());
    } else {
      this.logger = createLogger(this.name);
    }
    return this;
  }

  /** Registers a keypress listener for Ctrl+C that initiates instance shutdown. */
  // private registerExitListener(): Instance {
  //   const self = this;
  //   readline.emitKeypressEvents(process.stdin);
  //   process.stdin.on("keypress", async (_, { ctrl, name }) => {
  //     if (ctrl && name === "c") {
  //       this.logger.info("shutting down...");
  //       await this.executionManager.shutdown();
  //       process.exit(0);
  //     }
  //   });
  //   process.stdin.setRawMode(true);
  //   process.stdin.resume();
  //   return self;
  // }

  /** Registers a state listener that persists to disk. */
  private registerStateListener(): Instance {
    const set =
      (instance: Instance) =>
      (target: Queue, prop: string | symbol, receiver: never) => {
        // wait until reflected, then persist state
        setTimeout((instance) => instance.persistState(), 100, instance);
        return Reflect.set(target, prop, receiver);
      };

    new Proxy(this.executionManager.queue, {
      set: set(this),
    });

    return this;
  }

  private async persistState() {
    const serializedState = JSON.stringify({
      queue: this.executionManager.queue,
      xtx: this.executionManager.xtx,
      sfxToXtx: this.executionManager.sfxToXtx,
    });
    await writeFile(this.stateFile, serializedState);
  }
}

export {
  Instance,
  ExecutionManager,
  Queue,
  Execution,
  SideEffect,
  Notification,
  NotificationType,
  TxOutput,
  TxStatus,
  SubstrateRelayer,
  Estimator,
  CostEstimator,
  Estimate,
  InclusionProof,
  BiddingEngine,
  StrategyEngine,
  SfxStrategy,
  XtxStrategy,
  PriceEngine,
  CoingeckoPricing,
  CircuitListener,
  ListenerEvents,
  ListenerEventData,
  CircuitRelayer,
  Config,
  Circuit,
  Gateway,
  Strategy,
  BiddingStrategy,
};
