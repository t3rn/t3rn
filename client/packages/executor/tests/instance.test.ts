import { default as chai, expect } from "chai";
import chaiAsPromised from "chai-as-promised";
import { jestSnapshotPlugin } from "mocha-chai-jest-snapshot";
import { join } from "path";
import { homedir } from "os";
import { existsSync } from "fs";
import { readFile, readdir, mkdir, unlink } from "fs/promises";
import { Instance } from "../src";

chai.use(chaiAsPromised);
chai.use(jestSnapshotPlugin());
chai.should();

describe("Instance", () => {
  let name = "alina";

  describe("Configuration", () => {
    let conf = join(homedir(), `.t3rn-executor-${name}`, "config.json");
    let instance;

    beforeEach(async () => {
      if (existsSync(conf)) {
        await unlink(conf);
      }
      process.env.CIRCUIT_SIGNER_KEY = `0x${"acab".repeat(16)}`;
      process.env.ROCO_GATEWAY_SIGNER_KEY = `0x${"acab".repeat(16)}`;
      instance = new Instance(name, false /*logToDisk*/);
      instance.logger = { warn() { }, info() { } };
    });

    it("should throw if signer keys are missing", async () => {
      process.env.CIRCUIT_SIGNER_KEY = undefined;
      process.env.ROCO_GATEWAY_SIGNER_KEY = undefined;

      instance
        .loadConfig()
        .should.be.rejectedWith(
          Error,
          "Instance::loadConfig: missing circuit signer key"
        );
    });

    it("should throw if signer keys are malformatted", async () => {
      process.env.CIRCUIT_SIGNER_KEY = "acab";
      process.env.ROCO_GATEWAY_SIGNER_KEY = "acab";

      instance
        .loadConfig()
        .should.be.rejectedWith(
          Error,
          "Instance::loadConfig: missing circuit signer key"
        );

      // reset to bogus substrate private key for the remainder
      // somehow the beforeEach hook resets don't suffice
      process.env.CIRCUIT_SIGNER_KEY = `0x${"acab".repeat(16)}`;
      process.env.ROCO_GATEWAY_SIGNER_KEY = `0x${"acab".repeat(16)}`;
    });

    it("should load custom config", async () => {
      expect(instance.config).to.be.undefined;

      let config = await instance.loadConfig();

      expect(instance.config).to.not.be.undefined;
      expect(config).to.deep.equal(instance.config);
      expect(instance.config).toMatchSnapshot();
    });

    it("should persist custom config", async () => {
      expect(existsSync(conf)).to.be.false;

      let config = await instance.loadConfig();

      expect(existsSync(conf)).to.be.true;
      let stored = await readFile(conf, "utf8").then(JSON.parse);
      expect(stored).to.deep.equal(config);
      expect(stored).toMatchSnapshot();
    });
  });

  describe("Logs", () => {
    let logs = join(homedir(), `.t3rn-executor-${name}`, "logs");
    let instance;

    beforeEach(async () => {
      await mkdir(logs, { recursive: true });
      await readdir(logs).then((logFiles) =>
        Promise.all(logFiles.map((logFile) => unlink(join(logs, logFile))))
      );
      process.env.CIRCUIT_SIGNER_KEY = `0x${"acab".repeat(16)}`;
      process.env.ROCO_GATEWAY_SIGNER_KEY = `0x${"acab".repeat(16)}`;
      instance = new Instance(name);
    });

    it("should not log to disk", async () => {
      let logFiles = await readdir(logs);
      expect(logFiles.length).to.equal(0);

      await instance.configureLogging();
      instance.logger = { warn() { }, info() { } } as any;
      await instance.logger.info("hallo");

      logFiles = await readdir(logs);
      expect(logFiles.length).to.equal(0);
    });

    it("should log to disk", async () => {
      let instance = new Instance(name, true /**logToDisk*/);
      let logFiles = await readdir(logs);
      expect(logFiles.length).to.equal(0);

      await instance.configureLogging();
      await instance.logger.info("hallo");

      logFiles = await readdir(logs);
      expect(logFiles.length).to.equal(1);
      let logged = await readFile(join(logs, logFiles[0]), "utf8");
      expect(logged).to.match(/hallo/);
    });
  });
}).afterAll(() => {
  process.exit(0);
});
