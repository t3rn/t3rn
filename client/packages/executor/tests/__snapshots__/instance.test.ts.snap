// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Instance Configuration should load custom config 1`] = `
Object {
  "assets": Object {
    "BSX": Array [
      Object {
        "id": "basilisk",
        "priceSource": "coingecko",
      },
    ],
    "ROC": Array [
      Object {
        "id": "polkadot",
        "priceSource": "coingecko",
      },
    ],
    "TRN": Array [
      Object {
        "id": "kusama",
        "priceSource": "coingecko",
      },
    ],
  },
  "bidding": Object {
    "bidAggressive": true,
    "bidMeek": false,
    "bidPercentile": 0.75,
    "closerPercentageBid": 0.1,
    "equalMinProfitBid": false,
    "overrideNoCompetition": true,
  },
  "circuit": Object {
    "decimals": 12,
    "rpc": "wss://rpc.t0rn.io",
    "signerKey": "0xacabacabacabacabacabacabacabacabacabacabacabacabacabacabacabacab",
    "ticker": "TRN",
  },
  "default": Object {
    "assets": Object {
      "BSX": Array [
        Object {
          "id": "basilisk",
          "priceSource": "coingecko",
        },
      ],
      "ROC": Array [
        Object {
          "id": "polkadot",
          "priceSource": "coingecko",
        },
      ],
      "TRN": Array [
        Object {
          "id": "kusama",
          "priceSource": "coingecko",
        },
      ],
    },
    "bidding": Object {
      "bidAggressive": true,
      "bidMeek": false,
      "bidPercentile": 0.75,
      "closerPercentageBid": 0.1,
      "equalMinProfitBid": false,
      "overrideNoCompetition": true,
    },
    "circuit": Object {
      "decimals": 12,
      "rpc": "wss://rpc.t0rn.io",
      "signerKey": "0xacabacabacabacabacabacabacabacabacabacabacabacabacabacabacabacab",
      "ticker": "TRN",
    },
    "gateways": Array [
      Object {
        "id": "roco",
        "name": "Rococo",
        "rpc": "wss://rococo-rpc.polkadot.io",
        "signerKey": "0xacabacabacabacabacabacabacabacabacabacabacabacabacabacabacabacab",
        "type": "Substrate",
      },
      Object {
        "id": "bslk",
        "name": "Basilisk",
        "rpc": "wss://rpc-01.basilisk-rococo.hydradx.io",
        "type": "Substrate",
      },
    ],
    "pricing": Object {
      "coingecko": Object {
        "endpoint": "https://api.coingecko.com/api/v3/coins/",
      },
    },
    "strategies": Object {
      "roco": Object {
        "sfx": Object {
          "minProfitUsd": 3,
          "minYield": 0.05,
        },
        "supportedAssets": Array [
          "ROC",
        ],
        "xtx": Object {
          "minInsuranceAmountUsd": 1,
          "minInsuranceShare": 0.1,
        },
      },
    },
    "vendors": Array [
      "Rococo",
    ],
  },
  "gateways": Array [
    Object {
      "id": "roco",
      "name": "Rococo",
      "rpc": "wss://rococo-rpc.polkadot.io",
      "signerKey": "0xacabacabacabacabacabacabacabacabacabacabacabacabacabacabacabacab",
      "type": "Substrate",
    },
    Object {
      "id": "bslk",
      "name": "Basilisk",
      "rpc": "wss://rpc-01.basilisk-rococo.hydradx.io",
      "type": "Substrate",
    },
  ],
  "name": "alina",
  "pricing": Object {
    "coingecko": Object {
      "endpoint": "https://api.coingecko.com/api/v3/coins/",
    },
  },
  "strategies": Object {
    "roco": Object {
      "sfx": Object {
        "minProfitUsd": 3,
        "minYield": 0.05,
      },
      "supportedAssets": Array [
        "ROC",
      ],
      "xtx": Object {
        "minInsuranceAmountUsd": 1,
        "minInsuranceShare": 0.1,
      },
    },
  },
  "vendors": Array [
    "Rococo",
  ],
}
`;

exports[`Instance Configuration should persist custom config 1`] = `
Object {
  "assets": Object {
    "BSX": Array [
      Object {
        "id": "basilisk",
        "priceSource": "coingecko",
      },
    ],
    "ROC": Array [
      Object {
        "id": "polkadot",
        "priceSource": "coingecko",
      },
    ],
    "TRN": Array [
      Object {
        "id": "kusama",
        "priceSource": "coingecko",
      },
    ],
  },
  "bidding": Object {
    "bidAggressive": true,
    "bidMeek": false,
    "bidPercentile": 0.75,
    "closerPercentageBid": 0.1,
    "equalMinProfitBid": false,
    "overrideNoCompetition": true,
  },
  "circuit": Object {
    "decimals": 12,
    "rpc": "wss://rpc.t0rn.io",
    "signerKey": "0xacabacabacabacabacabacabacabacabacabacabacabacabacabacabacabacab",
    "ticker": "TRN",
  },
  "default": Object {
    "assets": Object {
      "BSX": Array [
        Object {
          "id": "basilisk",
          "priceSource": "coingecko",
        },
      ],
      "ROC": Array [
        Object {
          "id": "polkadot",
          "priceSource": "coingecko",
        },
      ],
      "TRN": Array [
        Object {
          "id": "kusama",
          "priceSource": "coingecko",
        },
      ],
    },
    "bidding": Object {
      "bidAggressive": true,
      "bidMeek": false,
      "bidPercentile": 0.75,
      "closerPercentageBid": 0.1,
      "equalMinProfitBid": false,
      "overrideNoCompetition": true,
    },
    "circuit": Object {
      "decimals": 12,
      "rpc": "wss://rpc.t0rn.io",
      "signerKey": "0xacabacabacabacabacabacabacabacabacabacabacabacabacabacabacabacab",
      "ticker": "TRN",
    },
    "gateways": Array [
      Object {
        "id": "roco",
        "name": "Rococo",
        "rpc": "wss://rococo-rpc.polkadot.io",
        "signerKey": "0xacabacabacabacabacabacabacabacabacabacabacabacabacabacabacabacab",
        "type": "Substrate",
      },
      Object {
        "id": "bslk",
        "name": "Basilisk",
        "rpc": "wss://rpc-01.basilisk-rococo.hydradx.io",
        "type": "Substrate",
      },
    ],
    "pricing": Object {
      "coingecko": Object {
        "endpoint": "https://api.coingecko.com/api/v3/coins/",
      },
    },
    "strategies": Object {
      "roco": Object {
        "sfx": Object {
          "minProfitUsd": 3,
          "minYield": 0.05,
        },
        "supportedAssets": Array [
          "ROC",
        ],
        "xtx": Object {
          "minInsuranceAmountUsd": 1,
          "minInsuranceShare": 0.1,
        },
      },
    },
    "vendors": Array [
      "Rococo",
    ],
  },
  "gateways": Array [
    Object {
      "id": "roco",
      "name": "Rococo",
      "rpc": "wss://rococo-rpc.polkadot.io",
      "signerKey": "0xacabacabacabacabacabacabacabacabacabacabacabacabacabacabacabacab",
      "type": "Substrate",
    },
    Object {
      "id": "bslk",
      "name": "Basilisk",
      "rpc": "wss://rpc-01.basilisk-rococo.hydradx.io",
      "type": "Substrate",
    },
  ],
  "name": "alina",
  "pricing": Object {
    "coingecko": Object {
      "endpoint": "https://api.coingecko.com/api/v3/coins/",
    },
  },
  "strategies": Object {
    "roco": Object {
      "sfx": Object {
        "minProfitUsd": 3,
        "minYield": 0.05,
      },
      "supportedAssets": Array [
        "ROC",
      ],
      "xtx": Object {
        "minInsuranceAmountUsd": 1,
        "minInsuranceShare": 0.1,
      },
    },
  },
  "vendors": Array [
    "Rococo",
  ],
}
`;
